### PR TITLE
Model Monitors endpoints

### PR DESCRIPTION
**Pull request by sinnec**

### CODE REVIEW

File: src/api/v1/model_monitors.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    if body is not None:
+        new_model_monitor = crud.model_monitors.create(db=db, obj_in=body)
+        return new_model_monitor
+    else:
+        return errors.bad_request("Body should not be empty")
+
+
+@model_monitors_router.get(
+    "/models/{model_id}/model-monitors",
+    tags=["Model Monitors"],
```
Comment (momegas) [inline src/api/v1/model_monitors.py:38]: If this starts with `/model/...` it should be under models tags

File: src/schemas/modelMonitor.py [hunk 2]
```diff
@@ -1,14 +1,37 @@
-from typing import Any, Optional, Union
 from pydantic import BaseModel
+import enum
 from src.schemas.base import ItemBase
 
 
+class MonitorStatus(str, enum.Enum):
+    open = "open"
+    closed = "closed"
```
Comment (momegas) [inline src/schemas/modelMonitor.py:8]: Active / inactive maybe?

File: src/schemas/modelMonitor.py [hunk 3]
```diff
@@ -1,14 +1,37 @@
...
     model_id: str
-    dataset_id: str
     name: str
-    # metric: Any #Enum
-    threshold: Union[int, float]
+    status: MonitorStatus
+    metric: MonitorMetrics
+    threshold: float
+    severity: AlertSeverity
+    email: str
```
Comment (momegas) [inline src/schemas/modelMonitor.py:34]: Some food for though on this is that the drifting in multiclass models is measured per class.
As is missing values count in general.

File: src/api/v1/model_monitors.py (hunk 1 above)
Comment (sinnec) [reply src/api/v1/model_monitors.py:38]: Well I was afraid that these endpoints (along with some previous ones) need a different path. I surely don't want to add them to the models Tag, so would `model-monitors/{model_id}` be a better naming?

File: src/schemas/modelMonitor.py (hunk 2 above)
Comment (sinnec) [reply src/schemas/modelMonitor.py:8]: Could be 😉

File: src/api/v1/model_monitors.py (hunk 1 above)
Comment (sinnec) [reply src/api/v1/model_monitors.py:38]: Or `{model_id}/model-monitors`?

File: .gitignore [hunk 4]
```diff
@@ -7,6 +7,6 @@ __pycache__/
 .pytest_cache
 *.DS_Store
 check_data/
-models/
+/models/
```
Comment (gcharis) [inline .gitignore:10]: I don't think it works like that
try just `models`?

File: src/api/v1/model_monitors.py (hunk 1 above)
Comment (gcharis) [reply src/api/v1/model_monitors.py:38]: Why not just `{model_id}/monitors`?

File: .gitignore (hunk 4 above)
Comment (sinnec) [reply .gitignore:10]: Well there are more than one folders named 'models' so I just want only the one that's in the root directory and keeps the trained models. My change seems to work.

File: src/api/v1/model_monitors.py (hunk 1 above)
Comment (sinnec) [reply src/api/v1/model_monitors.py:38]: Well that's a solution but we discussed with Megas and another possible solution is: `/model-monitors?model_id=1111111`

Comment (gcharis) [reply src/api/v1/model_monitors.py:38]: Then why is it under the /model path?

Comment (sinnec) [reply src/api/v1/model_monitors.py:38]: Because it's wrong! Needs fixing!

Comment (gcharis) [reply src/api/v1/model_monitors.py:38]: You seem a little bit aggressive

Comment (sinnec) [reply src/api/v1/model_monitors.py:38]: Maybe the exclamation marks give a wrong impression! 😋

File: src/api/v1/model_monitors.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    body: ModelMonitorCreateDto,
+    db: Session = Depends(get_db),
+    authenticated_user: User = Depends(authenticate_user),
+) -> ModelMonitor:
+
+    if body is not None:
+        new_model_monitor = crud.model_monitors.create(db=db, obj_in=body)
+        return new_model_monitor
+    else:
+        return errors.bad_request("Body should not be empty")
```
Comment (momegas) [inline src/api/v1/model_monitors.py:33]: I think the check here is not needed. Pydantic will catch this right?

File: src/cron_tasks/monitoring_alerts.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            for monitor in model_monitors:
+                if monitor.metric in [
+                    MonitorMetrics.accuracy,
+                    MonitorMetrics.precision,
+                    MonitorMetrics.recall,
+                    MonitorMetrics.f1,
+                ]:
+                    await run_create_performance_metric_alert_pipeline(model, monitor)
+                elif monitor.metric == MonitorMetrics.data_drift:
+                    await run_create_data_drift_alert_pipeline(model, monitor)
```
Comment (momegas) [inline src/cron_tasks/monitoring_alerts.py:96]: Can we skip the else here? You can just return on the `if not models` above.

**Review by momegas — APPROVED** [review]
This is an awesome first iteration.
Great Job 👍

File: src/api/v1/model_monitors.py (hunk 5 above)
Comment (sinnec) [reply src/api/v1/model_monitors.py:33]: Fixed!

File: src/cron_tasks/monitoring_alerts.py (hunk 6 above)
Comment (sinnec) [reply src/cron_tasks/monitoring_alerts.py:96]: I guess, but will need to repeat lines 98-101 within the `in not models` (closing db and all the logging)

File: src/cron_tasks/monitoring_alerts.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    last_performance_metrics_report = await get_latest_performance_metrics_report(
+        db, model
+    )
+
+    if model.type == ModelType.binary or monitor.metric == MonitorMetrics.accuracy:
+        metric_value = vars(last_performance_metrics_report)[monitor.metric]
+    else:
+        metric_value = vars(last_performance_metrics_report)[monitor.metric]["weighted"]
+
+    if metric_value < monitor.threshold:
```
Comment (gcharis) [inline src/cron_tasks/monitoring_alerts.py:42]: Should we abstract this comparison? 
@momegas

Comment (gcharis) [reply src/cron_tasks/monitoring_alerts.py:42]: @sinnec maybe we should rename the `threshold` to sth like `upper_threshold`

File: src/cron_tasks/monitoring_alerts.py (hunk 6 above)
Comment (gcharis) [reply src/cron_tasks/monitoring_alerts.py:96]: There is more code to be executed at the bottom

**Pull request merged**